lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -0,0 +1,245 @@
+    session_map[imsi] = std::vector<std::unique_ptr<SessionState>>();
+  } else {
+    MLOG(MINFO) << "Sanjay inside else session exist session for IMSI " << imsi;
+    session_map[imsi].push_back(std::move(session_state));
+  }
+  return true;
+}
+
+void SessionStateEnforcer::handle_session_init_rule_updates(
+    SessionMap& session_map, const std::string& imsi,
+    SessionState& session_state) {
+  auto itp = pdr_map_.equal_range(imsi);
+  for (auto itr = itp.first; itr != itp.second; itr++) {
+    /* Get the PDR numbers, now  get the rules from global static rule
+     * list
+     */
+    SetGroupPDR rule;
+    GlobalRuleList.get_rule(itr->second, &rule);
+    session_state.insert_with_static_rules(&rule);
+  }
+  auto itf = far_map_.equal_range(imsi);
+  for (auto itr = itf.first; itr != itf.second; itr++) {
+    /* Get the PDR numbers, now  get the rules from global static rule
+     * list
+     */
+    SetGroupFAR rule;
+    GlobalRuleList.get_rule(itr->second, &rule);
+    // Add to the the session vector
+    session_state.insert_with_static_rules(&rule);
+  }
+  auto ip_addr = session_state.get_config()
+                     .rat_specific_context.m5gsm_session_context()
+                     .pdu_address()
+                     .redirect_server_address();
+  SessionState::SessionInfo sess_info;
+  sess_info.imsi       = imsi;
+  sess_info.ip_addr    = ip_addr;
+  sess_info.Pdr_rules_ = session_state.get_5g_static_pdr_rules();
+  sess_info.Far_rules_ = session_state.get_5g_static_far_rules();
+  session_state.sess_infocopy(&sess_info);
+
+  /* session_state elments are filled with rules. State needs to be
+   * moved to CREATED and sending message to UPF.
+   * Note: charging and credit related info not taken care in drop-1
+   *
+   * TODO - will be taken care later
+   * Thinking of adding the created session into session store after step
+   * CREATING state and before rules are add to session. And adding the logic
+   * of implementing policy rules into the event loop. This way, when PCF
+   * component is introduced an asynchronous call into PCF (PolicyDB) can be
+   * handled easily later on.
+   *
+   */
+  auto update_criteria = get_default_update_criteria();
+  session_state.set_fsm_state(CREATED, update_criteria);
+  MLOG(MDEBUG) << "State of session changed to "
+               << session_fsm_state_to_str(session_state.get_state());
+  /* Update the m5gsm_cause and prepare for respone along with actual cause*/
+  prepare_response_to_access(
+      imsi, session_state, magma::lte::M5GSMCause::OPERATION_SUCCESS);
+  pipelined_client_->set_upf_session(sess_info, call_back_void_upf);
+  return;
+}
+
+/* To send response back to AMF
+ * Fill the response structure from session context message
+ * and call rpc of AmfServiceClient
+ * TODO  const std::vector<magma::PolicyRule>& flows
+ *            ==> related to AuthorizedQosRules authorized_qos_rules.
+ */
+void SessionStateEnforcer::prepare_response_to_access(

```

Comment:
shouldnt this function be part of `amf_srv_client` 
The structure of the response should be not seen at `SessionStarteEnforcer` level

Reply:
This function to prepare the response and then calling amf_srv_client function handle_response_to_access to send message to AMF.

Reply:
So to to follow the other calls we have in the system, can `handle_response_to_access` do both things inside the amf?  We should hide the message creation on SessionStateEnforcer

Something like:
```
  amf::sendAccessToAMF(configObject& config){
	auto req = createAccessRequest(config);
	handle_response_to_access(request);
}

```

Note we do the same with the other GRPC clients, where the client takes care of building the response.  That removes some code that does not include business logic from the Enforcers.

Reply:
This is response, not all elements of config are required to be sent and we need to build the structure before sending back.
I will add TODO to address this comment in new PR and change the code accordingly.
Note: The Handling request and sending response back is handled by same Enforcer main thread context. :)